### PR TITLE
DEV-4481: pyangbind xml decoder does not support restricted yang values types which base type is string

### PR DESCRIPTION
**Pull request by eduardpo**
## Description

pyangbind xml decoder fails on restricted yang values which corresponding bindings base type is string.

This seems to be an artifact of the 2 to 3 python conversion. In the RestrictedClassType in yangtypes.py, in the method mp_check there is a type check that looks for six types.

           def mp_check(value):
                if not isinstance(value, six.string_types + (six.text_type,)):
                    return False
                if regex.match(convert_regexp(regexp), value):
                    return True
                return False

            return mp_check
However if the type passed in is a Restricted type it is passed in as an lxml.objectify.StringElement.
This fails the check.
If instead of the lxml.objectify.StringElement you use the lxml.objectify.StringElement.pyval, it will work.

My solution was to add a type check before the call to mp_check.
in class RestrictedClass(base_type):

        if val is not False:
            for test in self._restriction_tests:
                passed = False
                if isinstance(val, objectify.StringElement):
                    val = val.pyval
                if test(val) is not False:
                    passed = True
                    break
            if not passed:
                raise ValueError("%s does not match a restricted type" % val)

## References
JIRA: 4481

### CODE REVIEW

**Comment by shmuelhazan** [comment]
I have not managed to understand the issue and the description. Can you rephrase the description? 
Thanks.

**Comment by eduardpo** [comment]
> I have not managed to understand the issue and the description. Can you rephrase the description? Thanks.
@shmuelhazan 
done, is it better?

**Comment by shmuelhazan** [comment]
> > I have not managed to understand the issue and the description. Can you rephrase the description? Thanks.
> > @shmuelhazan
> > done, is it better?

Hi Eduard, I still can't understand, can you add a few sentences that explain what was the issue and how it has been fixed? This is a non-trivial issue.

**Comment by eduardpo** [comment]
> > > I have not managed to understand the issue and the description. Can you rephrase the description? Thanks.
> > > @shmuelhazan
> > > done, is it better?
> 
> Hi Eduard, I still can't understand, can you add a few sentences that explain what was the issue and how it has been fixed? This is a non-trivial issue.

Hi Shmuel,
For discovery, reproduction steps and solution, please refer to the Jira issue: 
https://jira-siklu.atlassian.net/browse/DEV-4481

**Comment by eduardpo** [comment]
To reproduce the phenomenon described in DEV-4481,
It's enough to run recent commit of DEV-4330 which implementation involves YANG models that triggers the phenomenone in pyangbind issue.
The solution for this problem was kindly provided by **Ian Grush**:
https://github.com/robshakir/pyangbind/issues/244

**Summary:**
I have discovered the issue. This seems to be an artifact of the 2 to 3 python conversion. In the RestrictedClassType in yangtypes.py, in the method mp_check there is a type check that looks for six types.

               def mp_check(value):
                    if not isinstance(value, six.string_types + (six.text_type,)):
                        return False
                    if regex.match(convert_regexp(regexp), value):
                        return True
                    return False

                return mp_check
However if the type passed in is a Restricted type it is passed in as an lxml.objectify.StringElement.
This fails the check.
If instead of the lxml.objectify.StringElement you use the lxml.objectify.StringElement.pyval, it will work.

My solution was to add a type check before the call to mp_check.
in class RestrictedClass(base_type):

            if val is not False:
                for test in self._restriction_tests:
                    passed = False
                    if isinstance(val, objectify.StringElement):
                        val = val.pyval
                    if test(val) is not False:
                        passed = True
                        break
                if not passed:
                    raise ValueError("%s does not match a restricted type" % val)

**Comment by shmuelhazan** [comment]
@eduardpo Can you please edit the description instead of adding a comment?

**Review by shmuelhazan — APPROVED** [review]
